File: modules/core/src/main/scala/one/xingyi/utils/simpleServer/HttpUtils.scala
package one.xingyi.utils.simpleServer

import java.util.concurrent.{Callable, ExecutorService, Executors}

import com.sun.net.httpserver.HttpExchange
import one.xingyi.utils.http.{Body, ContentType, ServiceResponse, Status}

import scala.io.Source

object HttpUtils {
  def makeDefaultExecutor: ExecutorService = Executors.newFixedThreadPool(100)


  def write(exchange: HttpExchange, response: ServiceResponse): Unit = {
    exchange.getResponseHeaders.set("content-type", response.contentType.s)
    val bytes = response.body.s.getBytes("UTF-8")
    exchange.sendResponseHeaders(response.status.code, bytes.length)
    Streams.sendAll(exchange.getResponseBody, bytes)
  }

  def process(exchange: HttpExchange)(response: => Option[ServiceResponse]): Unit = {
    try {
      val result = response
      result match {
        case None => write(exchange, new ServiceResponse(Status(404), Body(s"not found. ${exchange.getRequestURI}"), ContentType("text/plain")))
        case Some(x) => write(exchange, x)
      }
    }
    catch {
      case e: Exception => write(exchange, new ServiceResponse(Status(500), Body(e.getClass.getName + "\n" + e.getMessage), ContentType("text/plain")))
    }
  }


  def id(httpExchange: HttpExchange, prefix: String): String = httpExchange.getRequestURI.getPath.substring(prefix.length + 1)
}

File: modules/javaserver/src/main/java/one/xingyi/javaserver/HttpUtils.java
package one.xingyi.javaserver;

import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class    HttpUtils {

    public static ExecutorService makeDefaultExecutor() {
        return Executors.newFixedThreadPool(100);
    }

    public static void write(HttpExchange exchange, SimpleHttpResponse response) throws IOException {
        exchange.getResponseHeaders().set("content-type", response.getContentType());
        byte[] bytes = response.getBody().getBytes("UTF-8");
        exchange.sendResponseHeaders(response.getStatusCode(), bytes.length);
        Streams.sendAll(exchange.getResponseBody(), bytes);
    }

    public static void process(HttpExchange exchange, Callable<SimpleHttpResponse> callable) throws IOException {
        try {
            HttpUtils.write(exchange, callable.call());
        } catch (Exception e) {
            HttpUtils.write(exchange, new SimpleHttpResponse(500, "text/plain", e.getClass().getName() + "\n" + e.getMessage()));
        }
    }


    public static String id(HttpExchange httpExchange, String prefix) {
        return httpExchange.getRequestURI().getPath().substring(prefix.length() + 1);
    }

}
